File: src/main/java/io/ryecrow/joe/util/FiletimeUtils.java
package io.ryecrow.joe.util;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.Month;
import java.time.ZoneId;
import java.time.temporal.ChronoUnit;

/**
 * Utility class for converting between Win32 FILETIME and {@link java.time.Instant}
 *
 * @author ryecrow
 * @since 1.0
 */
public final class FiletimeUtils {

    private static final long GAP_FROM_UNIX_EPOCH = -11644473600L;
    private static final Instant FILETIME_EPOCH = LocalDateTime.of(1601, Month.JANUARY, 1, 0, 0, 0, 0)
            .atZone(ZoneId.systemDefault()).toInstant();

    private FiletimeUtils() {
        throw new AssertionError("No instance of FiletimeUtils for you!");
    }

    public static Instant toInstant(byte[] filetime) {
        if (filetime == null) {
            throw new IllegalArgumentException("FILETIME must not be null");
        }
        long time = filetime[filetime.length - 1] & 0xff;
        long tmp;
        for (int index = filetime.length - 2; index >= 0; index--) {
            time = time << 8;
            tmp = (long) filetime[index] & 0xff;
            time |= tmp;
        }
        return toInstant(time);
    }

    public static Instant toInstant(long filetime) {
        // Assume file must be created after 1 January 1601
        if (filetime < 0) {
            throw new IllegalArgumentException("FILETIME must be positive");
        }
        long nano = (filetime % 10_000_000) * 100;
        long second = (filetime / 10_000_000) + GAP_FROM_UNIX_EPOCH;
        return Instant.ofEpochSecond(second, nano);
    }

    public static long fromInstant(Instant instant) {
        if (instant == null) {
            return 0L;
        }
        long seconds = FILETIME_EPOCH.until(instant, ChronoUnit.SECONDS);
        long nano = instant.getNano();
        return (seconds * 10_000_000) + (nano / 100);
    }
}

File: src/main/kotlin/io/ryecrow/joe/util/FiletimeUtils.kt
package io.ryecrow.joe.util

import java.time.Instant
import java.time.LocalDateTime
import java.time.Month
import java.time.ZoneId
import java.time.temporal.ChronoUnit

/* Gap between 1 Jan 1601 and 1 Jan 1970, in seconds */
private const val GAP_FROM_UNIX_EPOCH = -11644473600L
private val FILETIME_EPOCH = LocalDateTime.of(1601, Month.JANUARY, 1, 0, 0, 0, 0)
    .atZone(ZoneId.systemDefault()).toInstant()

fun filetimeToInstant(filetime: ByteArray): Instant {
    var time: Long = (filetime[filetime.size - 1].toInt() and 0xff).toLong()
    var tmp: Long
    for (index in filetime.size - 2 downTo 0) {
        time = time shl 8
        tmp = filetime[index].toLong() and 0xff
        time = time or tmp
    }
    return filetimeToInstant(time)
}

fun filetimeToInstant(filetime: Long): Instant {
    // Assume file must be created after 1 January 1601
    require(filetime >= 0) { "FILETIME must be positive" }
    val nano = filetime % 10000000 * 100
    val second = filetime / 10000000 + GAP_FROM_UNIX_EPOCH
    return Instant.ofEpochSecond(second, nano)
}

fun Instant?.toFiletime(): Long {
    return if (this == null) {
        return 0L
    } else {
        val seconds = FILETIME_EPOCH.until(this, ChronoUnit.SECONDS)
        val nano = this.nano.toLong()
        seconds * 10000000 + nano / 100
    }
}